PAY-5710: First iteration of the Payment Services drop-in documentation (#162)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/payment-services/containers/credit-card.mdx

---
title: CreditCard container
description: Configure the CreditCard container to manage and display the available credit card fields during a checkout.
---

import OptionsTable from '@components/OptionsTable.astro';

> The `CreditCard` container is designed to allow placing an order with credit card details during the checkout process. You can configure it to manage and display the available credit card fields for a payment method.

Reviewer: ```suggestion
The `CreditCard` container allows a shopper to place an order with credit card details during the checkout process. You can configure it to manage and display the available credit card fields for a payment method.
```

## CreditCard configurations

The `CreditCard` container provides the following configuration options:

<OptionsTable
  compact
  options={[
    ['Option', 'Type', 'Req?', 'Description'],

>     ['apiUrl', 'string', 'Yes', 'Adobe Commerce GraphQL API URL, e.g., "https://magento.test/graphql".'],

Reviewer: ```suggestion
    ['apiUrl', 'string', 'Yes', 'The URL to the Adobe Commerce GraphQL endpoint, such as "https://example.com/graphql".'],
```

>     ['getCartId', 'function', 'Yes', 'Should return a promise that resolves to the cart id for which to make a payment.'],

Reviewer: ```suggestion
    ['getCartId', 'function', 'Yes', 'Should return a promise that resolves to the shopper's cart ID.'],
```

>     ['getCustomerToken', 'function', 'No', 'The `CreditCard` container may send GraphQL requests on behalf of the shopper. This requires GraphQL authorization, which can be performed using authorization tokens or session cookies. For token-based authorization, the "getCustomerToken" function that should return a customer token as a string, or null for guest checkouts.'],

Reviewer 2: Nested markdown is not supported, it seems.

![image](https://github.com/user-attachments/assets/3ed6f3d0-4073-494b-995d-7eec37ac5038)

Can we style this in some other way? If not, let's remove the backticks or use quotes.

Reviewer 2: Also, we should mention what we should pass in here for session-based authorization. We could go for something like this:

> [...] which can be performed using authorization tokens or session cookies. The "getCustomerToken" function should not be provided for session-based authorization. For token-based authorization, on the other hand, the "getCustomerToken" function is required and should return a customer token as a string or null for guest checkouts.

>     ['getCustomerToken', 'function', 'No', 'The `CreditCard` container may send GraphQL requests on behalf of the shopper. This requires GraphQL authorization, which can be performed using authorization tokens or session cookies. For token-based authorization, the "getCustomerToken" function that should return a customer token as a string, or null for guest checkouts.'],

Reviewer 2: Lastly, `Req?: No` might not be entirely correct. Strictly speaking, this parameter is conditionally required. 

* Merchant uses session-based GraphQL auth → `getCustomerToken` must not be provided (or must be `undefined`).

* Merchant uses token-based GraphQL auth → `getCustomerToken` must be provided and must return the customer token as a string or `null` if there is no customer logged in (i.e., guest).

>     ['getCustomerToken', 'function', 'No', 'The Credit Card container may send GraphQL requests on behalf of the customer. This requires GraphQL Authorization, which can be done in two ways: (1) Authorization tokens; (2) Session cookies. For token-based authorization, one must provide a "getCustomerToken" function that should return a customer token as a string or null for guest checkouts.'],

Reviewer: ```suggestion
    ['getCustomerToken', 'function', 'No', 'The `CreditCard` container may send GraphQL requests on behalf of the shopper. This requires GraphQL authorization, which can be performed using authorization tokens or session cookies. For token-based authorization, the "getCustomerToken" function that should return a customer token as a string, or null for guest checkouts.'],
```
What about session cookies?

>     ['getCustomerToken', 'function', 'Yes', 'The credit card container may send GraphQL requests on behalf of the shopper. This requires GraphQL authorization, which can be performed using authorization tokens or session cookies. For token-based authorization, the "getCustomerToken" function that should return a customer token as a string, or null for guest checkouts. The "getCustomerToken" function should not be provided for session-based authorization.'],

Reviewer: ```suggestion
    ['getCustomerToken', 'function', 'Yes', 'The credit card container may send GraphQL requests on behalf of the shopper. This requires GraphQL authorization, which can be performed using authorization tokens or session cookies. For token-based authorization, the "getCustomerToken" function should return a customer token as a string, or null for guest checkouts. The "getCustomerToken" function should not be provided for session-based authorization.'],
```

Reviewer: I don't think this should be marked as required, since it shouldn't be used for session authentication.

    ['onSuccess', 'function', 'No', 'Called when payment flow finishes successfully. This function is not passed any arguments.'],
    ['onError', 'function', 'No', 'Called when the payment flow was aborted due to an error. The function receives a single argument, {message: string}, containing the reason of error.'],
    ['onRender', 'function', 'No', 'Called when the credit card fields render successfully. The function receives the credit card instance as its single argument, which has a submit() method to be used to start the payment flow.'],

>     ['onValidation', 'function', 'No', 'Called when the credit card form validity changes. Use this, for example, to disable/enable a "Place order" button.'],

Reviewer: ```suggestion
    ['onValidation', 'function', 'No', 'Called when the credit card form validity changes. Use this to disable/enable a "Place order" button.'],
```

  ]}
/>

## Example

The following example renders the `CreditCard` container to handle and display the available credit card fields as a payment method.

```js
import CreditCard, { CREDIT_CARD_CODE } from '@dropins/payment-services/containers/CreditCard.js';
import { render as PaymentServicesProvider } from '@dropins/payment-services/render.js';
import { getConfigValue } from '../../scripts/configs.js';
import { getUserTokenCookie } from '../../scripts/initializers/index.js';

const apiUrl = await getConfigValue('commerce-core-endpoint');
const container = document.createElement('div');
const placeOrderButton = document.getElementById('place-order');
const getCartId = () => 'ozGi7uLI74etDYyMijoI2cla5CmGIBch';

PaymentServicesProvider.render(CreditCard, {
  apiUrl: apiUrl,

>   getCustomerToken: getUserTokenCookie,

Reviewer 2: Before merging, we should make sure that `getUserTokenCookie` actually returns `null` for guest checkouts. If not, we can provide a lambda, à la `() => async getUserTokenCookie() || null`.
